### PR TITLE
Add Domain Role Check for Users

### PR DESCRIPTION
**Pull request by HemanthDogiparthi12**
Paused Merge/Close until confirmation from the user that https://github.com/MyPureCloud/terraform-provider-genesyscloud/pull/514 worked.

### CODE REVIEW

**Review by charliecon — APPROVED** [review]

File: genesyscloud/user_group_role_common.go [hunk 1]
```diff
@@ -77,6 +77,33 @@ func readSubjectRoles(subjectID string, authAPI *platformclientv2.AuthorizationA
...
+	var roles []string
+
+	data, response, err := userAPI.GetUserRoles(subjectID)
+	if err != nil {
+			return nil, response, diag.Errorf("Failed to get Roles for subject %s: %s", subjectID, err)
+	}
+
+	if data != nil && data.Roles != nil {
+			for _, role := range *data.Roles {
+					if role.Id != nil {
```
[the remark this answers is not in the document]
Comment (HemanthDogiparthi12) [reply genesyscloud/user_group_role_common.go:90]: data != nil might still be required. since we are doing a range on data.roles. !!

**Review by carnellj-genesys — APPROVED** [review]

File: genesyscloud/user_group_role_common.go [hunk 2]
```diff
@@ -247,3 +247,64 @@ func validateResourceRole(resourceName string, roleResourceName string, division
...
```
Comment (carnellj-genesys) [inline genesyscloud/user_group_role_common.go:295]: roleId will never be null here.  Also, you should never have a data.Roles as nil.  It is as defined as an array and the SDK will translate the data as an empty array

File: genesyscloud/user_group_role_common.go [hunk 3]
```diff
@@ -247,3 +247,64 @@ func validateResourceRole(resourceName string, roleResourceName string, division
...
+							roles = append(roles, *role.Id)
+					}
+			}
+	}
+
+	if data != nil && data.UnusedRoles != nil {
+			for _, role := range *data.UnusedRoles {
+					if role.Id != nil {
```
Comment (carnellj-genesys) [inline genesyscloud/user_group_role_common.go:303]: See earlier comments

File: genesyscloud/user_group_role_common.go [hunk 4]
```diff
@@ -247,3 +247,64 @@ func validateResourceRole(resourceName string, roleResourceName string, division
...
+}
+
+func getDomainIdRoles(subjectID string, userAPI *platformclientv2.UsersApi) ([]string, *platformclientv2.APIResponse, diag.Diagnostics) {
+	var roles []string
+
+	data, response, err := userAPI.GetUserRoles(subjectID)
+	if err != nil {
+			return nil, response, diag.Errorf("Failed to get Roles for subject %s: %s", subjectID, err)
+	}
+
```
Comment (carnellj-genesys) [inline genesyscloud/user_group_role_common.go:292]: You should check the response code for a 404.  If its not 404 the data object will never be nil.

Comment (HemanthDogiparthi12) [reply genesyscloud/user_group_role_common.go:295]: done.

**Comment by charliecon** [comment]
@HemanthDogiparthi12 Where are we with this PR? I see a lot of debugging lines still in it. Do these changes fix the bug or was it resolved elsewhere?

**Comment by HemanthDogiparthi12** [comment]
> @HemanthDogiparthi12 Where are we with this PR? I see a lot of debugging lines still in it. Do these changes fix the bug or was it resolved elsewhere?

Hey Charlie. I have added a time delay for this problem https://github.com/MyPureCloud/terraform-provider-genesyscloud/pull/514.  We are suspecting it could be a caching issue . Just waiting for the next release of Terraform to be published so that I can ask the client to try reproduce the issue. and if he is still facing the problem.

**Comment by carnellj-genesys** [comment]
@HemanthDogiparthi12 You need to merge this in order for it to get into the release

**Comment by HemanthDogiparthi12** [comment]
> @HemanthDogiparthi12 You need to merge this in order for it to get into the release

Hi John.. There is another change wrt Time Delay https://github.com/MyPureCloud/terraform-provider-genesyscloud/pull/504 . This change is merged and is good to go for the release. Once the release is done, will connect to the user and check if the caching problem is still popping up.  We can address this PR accordingly after that.

**Comment by HemanthDogiparthi12** [comment]
Bug fix done https://github.com/MyPureCloud/terraform-provider-genesyscloud/pull/504
This can be closed

**Pull request closed without merging**